### PR TITLE
Stock part datums: Creates a framework for stock parts to not physically exist in the world

### PR DESCRIPTION
**Pull request by Mothblocks**
Currently, every stock part is created inside every machine on initialize. This is costing us about 300ms. This was likely done because a lot of similar code that I've been ripping out recently has been related to the singularity being a common source of destruction, so everything created its contents to be able to move later to reduce lag.

Unfortunately, machines are a lot more complicated than paper bins or decks of cards. They need to be able to exchange parts, but also the following pattern is very common:

```dm
// component_parts is sometimes, incorrectly, contents
for (var/obj/item/stock_parts/capacitor/C in component_parts)
```

That means the parts have to exist *somewhere*. I played with creating a single one and passing it immutably around component_parts, but it was a nightmare.

---

Instead, this PR creates `/datum/stock_part`, a singleton instance that maps to a specific part and tier. Machines specify their requirements using these stock part datums, which are not initialized, but should function exactly like normal stock parts.

Interface looks like this:

```dm
/datum/stock_part/scanning_module
	tier = 1
	physical_object_type = /obj/item/stock_parts/scanning_module
	physical_object_base_type = /obj/item/stock_parts/scanning_module

/datum/stock_part/scanning_module/tier2
	tier = 2
	physical_object_type = /obj/item/stock_parts/scanning_module/adv

/datum/stock_part/scanning_module/tier3
	tier = 3
	physical_object_type = /obj/item/stock_parts/scanning_module/phasic

/datum/stock_part/scanning_module/tier4
	tier = 4
	physical_object_type = /obj/item/stock_parts/scanning_module/triphasic
```

And to use:

```patch
- for (var/obj/item/stock_parts/capacitor/C in component_parts)
+ for (var/datum/stock_part/capacitor/C in component_parts)
```

```patch
req_components = list(
    /obj/item/stack/glass = 2,
-    /obj/item/stock_parts/capacitor = 1,
+    /datum/stock_part/capacitor = 1,
)
```

This PR currently only updates scanning modules. There are a very small handful of machines that actually use these, which makes them a good starting point for focusing on the system itself. After this PR is confirmed to be working and merged, I'll make PRs for every other core type.

Also, unlike paper bins, we do not hold onto the part that was put into the machine, it is completely deleted. This means that if the stock part has any unique qualities for whatever reason, they won't be preserved. Only thing I can think of is RPG stats but the code is very much made more complex with that support.

---

To maintainers: Hide whitespace mode recommended

image

:cl:
fix: Fixed gulag teleporters not containing the correct parts.
/:cl:

### CODE REVIEW

File: code/game/machinery/constructable_frame.dm [hunk 1]
```diff
@@ -42,7 +42,7 @@
 		return
 
 	var/list/nice_list = list()
-	for(var/atom/component as anything in req_components)
+	for(var/component as anything in req_components)
```
Comment (JohnFulpWillard) [inline code/game/machinery/constructable_frame.dm:45]:
```suggestion
	for(var/component in req_components)
```

File: code/game/machinery/constructable_frame.dm [hunk 2]
```diff
@@ -63,20 +63,24 @@
 		return
 
 	req_component_names = list()
-	for(var/atom/component_path as anything in req_components)
+	for(var/component_path as anything in req_components)
```
Comment (JohnFulpWillard) [inline code/game/machinery/constructable_frame.dm:66]:
```suggestion
	for(var/component_path in req_components)
```

File: code/game/machinery/constructable_frame.dm [hunk 3]
```diff
@@ -281,41 +288,90 @@
...
+					continue
+
+				if (req_components[stock_part_path] == 0)
+					continue
+
+				if (!istype(P, stock_part_path))
+					continue
+
+				if(isstack(P))
+					var/obj/item/stack/S = P
```
Comment (JohnFulpWillard) [inline code/game/machinery/constructable_frame.dm:310]: single letter var ☹️

File: code/game/objects/items/circuitboards/circuitboard.dm [hunk 4]
```diff
@@ -106,18 +112,17 @@ micro-manipulator, console screen, beaker, Microlaser, matter bin, power cells.
 		return .
 
 	var/list/nice_list = list()
-	for(var/atom/component_path as anything in req_components)
+	for(var/component_path as anything in req_components)
```
Comment (JohnFulpWillard) [inline code/game/objects/items/circuitboards/circuitboard.dm:115]:
```suggestion
	for(var/component_path in req_components)
```

File: code/game/machinery/constructable_frame.dm (hunk 3 above)
Comment (Mothblocks) [reply code/game/machinery/constructable_frame.dm:310]: Not mine!

**Comment by Jackal-boop** [comment]
I can only imagine the torment in mothblocks mind as they optimize shit code

**Comment by Mothblocks** [comment]
I'm mostly distracting myself

**Comment by Mothblocks** [comment]
Other tests are infinite looping O_o

File: code/game/machinery/constructable_frame.dm (hunk 3 above)
Comment (MTandi) [reply code/game/machinery/constructable_frame.dm:310]: Nerd!
image

Comment (Mothblocks) [reply code/game/machinery/constructable_frame.dm:310]: When you're already touching the diff, you should also change the variable name. You'll notice I did the same.

When you're just adding code, changing all the variable names makes for a harder to read diff, and it's okay to let it aside unless a maintainer requires it.

Comment (jlsnow301) [reply code/game/machinery/constructable_frame.dm:310]: Be the change you want to see!

Comment (Mothblocks) [reply code/game/machinery/constructable_frame.dm:310]: The change I want to see is PR diffs that are easy to read, this proc is huge :)

**Pull request merged**